INTERLOK-3497 Support ConfigurationReporter in parent gradle

quotidian-ennui opened the pull request:
## Motivation

- INTERLOK-3287 (https://github.com/adaptris/interlok/pull/546) introduced a new ConfigDeprecation annotation which is more formal than relying on the "developer" to log warnings about deprecated components.
- INTERLOK-3497 (https://github.com/adaptris/interlok/pull/579) added in the capability for a report file to be generated in the correct format if various system properties are set.

This PR changes the parent build.gradle to build on those two changes.

## Modification

Use `org.gradle.util.VersionNumber.parse()` to do tests against the interlokVersion. Switch to using https://github.com/adaptris/interlok/pull/579 implementations when running -configtest if version >=3.12. _VersionNumber_ works well enough with our current versioning convention

```
import org.gradle.util.VersionNumber
 
def checkVersion(version) {
  return VersionNumber.parse( version ) >= VersionNumber.parse( '3.12-SNAPSHOT' )
}
 
task checkVersion {
 
  doLast {
    def versions = ["3.9-SNAPSHOT", "3.9.2-RELEASE", "3.9.3-RELEASE", "3.9.3.1-RELEASE", "3.9.3.2-RELEASE", "3.10-SNAPSHOT", "3.10.0-RELEASE", "3.10.1-RELEASE", "3.10.2-RELEASE", "3.11-SNAPSHOT-RELEASE", "3.11.0-RELEASE", "3.11.1-RELEASE", "3.12-SNAPSHOT", "3.12.0-RELEASE", "3.12.1-RELEASE", "4.0.0"] as String[]
    versions.each {
      println "${it} " + checkVersion("${it}")
    }
  }
}
...
> Task :checkVersion
3.9-SNAPSHOT false
3.9.2-RELEASE false
3.9.3-RELEASE false
3.9.3.1-RELEASE false
3.9.3.2-RELEASE false
3.10-SNAPSHOT false
3.10.0-RELEASE false
3.10.1-RELEASE false
3.10.2-RELEASE false
3.11-SNAPSHOT-RELEASE false
3.11.0-RELEASE false
3.11.1-RELEASE false
3.12-SNAPSHOT true
3.12.0-RELEASE false
3.12.1-RELEASE true
4.0.0 true
```

- Added additional helpers for handling report verification etc to avoid duplication of code.

## Result

If the version >= 3.12-SNAPSHOT then we switch to asking -configcheck to generate the report _for us_ so that verifyReport can do work.

~~A side effect of the change ( to avoid code duplication but keep feature parity) is that we now use the report file itself during the reporting phase which means that we see `CODE_SMELL,INFO`...~~

## Testing

If we take the configuration from https://github.com/adaptris/interlok/pull/579 and we have the build.gradle file

```
ext {
  interlokParentGradle = "https://raw.githubusercontent.com/adaptris-labs/interlok-build-parent/INTERLOK-3497-verify-config-check/build.gradle"
  interlokVersion = '3.11-SNAPSHOT'
}

allprojects {
  apply from: "${interlokParentGradle}"
}

configurations.all {
  resolutionStrategy.cacheChangingModulesFor 0, "seconds"
}

dependencies {
  // These dependencies probably don't matter.
  interlokRuntime ("org.fusesource.jansi:jansi:2.0.1")
}
```

Then we get the relevant output via interlokVerifyLog4j

```
> Task :interlokVerifyLog4j
... // content skipped
> Task :interlokVerify

Configuration warnings [3]:
Configuration warnings [3]:
[AlwaysFailService] is deprecated, use [com.adaptris.core.services.exception.ThrowExceptionService] instead
NullMessageConsumer uses destination, it has no meaning
[Adapter(MyInterlokInstance)] has a MessageErrorHandler with no behaviour; messages may be discarded upon exception
```

If we switch the version to be __3.12-SNAPSHOT__ then we get more meaningful output based on the changes from https://github.com/adaptris/interlok/pull/546 

```
> Task :interlokVerifyConfigCheck
... // logging skipped
> Task :interlokVerify

Configuration warnings [3]:
Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].consumer.destination][NullMessageConsumer]: Is deprecated. It will be removed in 4.0.0
Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[0]][AlwaysFailService(always-fail-cos)]: Is deprecated. It will be removed in a future version
Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[1].services[0]][AlwaysFailService]: Is deprecated. It will be removed in a future version
```

quotidian-ennui committed:
INTERLOK-3497 Support ConfigurationReporter in parent gradle

Based on VersionNumber.parse() switch between log4j parsing and
ConfigReporter implementations when running -configtest.
Add new helpers within the script for handling verify reports

quotidian-ennui commented:
@mcwarman I've tried to limit the changes so that you can "copy pasta" it into the internal build.gradle; not sure how successful I was.

sebastien-belin-adp reviewed (approved):
Works the way it's supposed to

quotidian-ennui committed:
INTERLOK-3497 Add filtering of CODE_SMELL from logging output

- Add new project properties for category + level
- Filter out the category + level from the report output for
full backwards compatibility behaviour

quotidian-ennui commented:
https://github.com/adaptris/interlok/pull/581 breaks this now, so waiting on that to be merged so we can change the properties...

quotidian-ennui committed:
INTERLOK-3497 Add the new props for warnings.

adaptris/interlok#581 changed the properties; so we add the additional properties
to, we can remove the interlok.verify.deprecated.* properties post 3.12.0

quotidian-ennui committed:
INTERLOK-3497 removed 'old' properties

3.12-SNAPSHOT(2020-12-09:10:57:22 GMT) switches to
interlok.verify.warning.filename etc; so we can remove the
intelrok.verify.deprecated variants.

quotidian-ennui commented:
You can add in unused shared components as per : https://github.com/adaptris/interlok/pull/581 and you now get
```
$ cat build/reports/interlok/verify/report.txt
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[0]][AlwaysFailService(always-fail-cos)]: Is deprecated. It will be removed in a future version
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[1].services[0]][AlwaysFailService]: Is deprecated. It will be removed in a future version
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].consumer.destination][NullMessageConsumer]: Is deprecated. It will be removed in 4.0.0
CODE_SMELL,INFO,Shared connection unused: unused-null-connection
CODE_SMELL,INFO,Shared service unused: shared-service-1
CODE_SMELL,INFO,Shared service unused: shared-service-2
CODE_SMELL,INFO,Shared service unused: shared-service-3
```

Requires `Bootstrap of Interlok 3.12-SNAPSHOT(2020-12-09:10:57:22 GMT) complete` or later...

mcwarman reviewed (approved):
```text
$ cat ./build/reports/interlok/verify/report.txt
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[1].services[0]][AlwaysFailService]: Is deprecated. It will be removed in a future version
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].serviceCollection.services[0]][AlwaysFailService(always-fail-cos)]: Is deprecated. It will be removed in a future version
CODE_SMELL,INFO,Interlok Deprecation Warning: [channelList.channels[0].workflowList.workflows[0].consumer.destination][NullMessageConsumer]: Is deprecated. It will be removed in 4.0.0
CODE_SMELL,INFO,Shared connection unused: unused-null-connection
CODE_SMELL,INFO,Shared service unused: shared-service-1
CODE_SMELL,INFO,Shared service unused: shared-service-2
CODE_SMELL,INFO,Shared service unused: shared-service-3
```
